nebula_examples/src/hesai/hesai_ros_offline_extract_bag_pcd.cpp

```diff
@@ -328,39 +313,45 @@ Status HesaiRosOfflineExtractBag::ReadBag()
         std::cout << "Found data in topic " << bag_message->topic_name << ": "
                   << bag_message->time_stamp << std::endl;
 
-        auto pointcloud_ts = driver_ptr_->ConvertScanToPointcloud(
-          std::make_shared<pandar_msgs::msg::PandarScan>(extracted_msg));
-        auto pointcloud = std::get<0>(pointcloud_ts);
-        auto fn = std::to_string(bag_message->time_stamp) + ".pcd";
-        //        pcl::io::savePCDFileBinary((o_dir / fn).string(), *pointcloud);
+        for (auto & pkt : extracted_msg.packets) {
+          auto pointcloud_ts = driver_ptr_->ParseCloudPacket(std::vector<uint8_t>(pkt.data.begin(), std::next(pkt.data.begin(), pkt.size)));
+          auto pointcloud = std::get<0>(pointcloud_ts);
 
-        if (needs_open) {
-          const rosbag2_storage::StorageOptions storage_options_w(
-            {(o_dir / std::to_string(bag_message->time_stamp)).string(), "sqlite3"});
-          const rosbag2_cpp::ConverterOptions converter_options_w(
-            {rmw_get_serialization_format(), rmw_get_serialization_format()});
-          writer_ = std::make_unique<rosbag2_cpp::writers::SequentialWriter>();
-          writer_->open(storage_options_w, converter_options_w);
-          writer_->create_topic(
-            {bag_message->topic_name, "pandar_msgs/msg/PandarScan", rmw_get_serialization_format(),
-              ""});
-          needs_open = false;
-        }
-        writer_->write(bag_message);
-        cnt++;
-        if (skip_num < cnt) {
-          out_cnt++;
-          if (only_xyz) {
-            pcl::PointCloud<pcl::PointXYZ> cloud_xyz;
-            pcl::copyPointCloud(*pointcloud, cloud_xyz);
-            writer.writeBinary((o_dir / fn).string(), cloud_xyz);
-          } else {
-            writer.writeBinary((o_dir / fn).string(), *pointcloud);
+          if (!pointcloud) {
+            continue;
+          }
+
+          auto fn = std::to_string(bag_message->time_stamp) + ".pcd";
+          //        pcl::io::savePCDFileBinary((o_dir / fn).string(), *pointcloud);
+
+          if (needs_open) {
+            const rosbag2_storage::StorageOptions storage_options_w(
+              {(o_dir / std::to_string(bag_message->time_stamp)).string(), "sqlite3"});
+            const rosbag2_cpp::ConverterOptions converter_options_w(
+              {rmw_get_serialization_format(), rmw_get_serialization_format()});
+            writer_ = std::make_unique<rosbag2_cpp::writers::SequentialWriter>();
+            writer_->open(storage_options_w, converter_options_w);
+            writer_->create_topic(
+              {bag_message->topic_name, "pandar_msgs/msg/PandarScan", rmw_get_serialization_format(),
```

Comment:
While we are breaking stuff, should we write Nebula packets?

Reply:
Yes. will do :+1:

Reply:
Addressed in [49d0664](https://github.com/tier4/nebula/pull/127/commits/49d06642b2613d2288c98c149ac6c985995f22ae) :+1: